File: LeetCode/Python/src/Remove_Nth_Node_From_End_Of_List.py
"""
@author - Anirudh Sharma
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def printList(node: ListNode):
    while node is not None:
        print(str(node.val), end=" ")
        node = node.next
    print()


def removeNthFromEnd(head: ListNode, n: int) -> ListNode:
    # Two pointers - fast and slow
    slow = head
    fast = head
    # Move fast pointer n steps ahead
    for i in range(0, n):
        if fast.next is None:
            # If n is equal to the number of nodes, delete the head node
            if i == n - 1:
                head = head.next
            return head
        fast = fast.next
    # Loop until fast node reaches to the end
    # Now we will move both slow and fast pointers
    while fast.next is not None:
        slow = slow.next
        fast = fast.next
    # Delink the nth node from last
    if slow.next is not None:
        slow.next = slow.next.next
    return head


if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)
    printList(removeNthFromEnd(head, 2))

    head = ListNode(1)
    printList(removeNthFromEnd(head, 1))

    head = ListNode(1)
    head.next = ListNode(2)
    printList(removeNthFromEnd(head, 1))

File: LeetCode/JavaScript/src/remove_nth_node_from_end_of_list.js
var removeNthFromEnd = function (head, n) {
    // Two pointers - fast and slow
    let slow = head;
    let fast = head;
    // Move fast pointer n steps ahead
    for (let i = 0; i < n; i++) {
        if (fast.next === null) {
            // If n is equal to the number of nodes, delete the head node
            if (i === n - 1) {
                head = head.next;
            }
            return head;
        }
        fast = fast.next;
    }
    // Loop until we reach to the end.
    // Now we will move both fast and slow pointers
    while (fast.next !== null) {
        slow = slow.next;
        fast = fast.next;
    }
    // Delink the nth node from last
    if (slow.next !== null) {
        slow.next = slow.next.next;
    }
    return head;
};

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

function printList(node) {
    let list = [];
    while (node != null) {
        list.push(node.val);
        node = node.next
    }
    console.log(list.join(" "));
}

head = new ListNode(1);
head.next = new ListNode(2);
head.next.next = new ListNode(3);
head.next.next.next = new ListNode(4);
head.next.next.next.next = new ListNode(5);
printList(removeNthFromEnd(head, 2));

head = new ListNode(1);
printList(removeNthFromEnd(head, 1));

head = new ListNode(1);
head.next = new ListNode(2);
printList(removeNthFromEnd(head, 1));

